### ISSUE TITLE
How we can make Intellij-Rust open source process better?

### ISSUE THREAD

**Issue by vlad20012** [intellij-rust/intellij-rust#1698]
Dear contributors!
You probably know that Intellij-Rust was started as truly open source/community driven project. And we want this project to continue to develop in the same manner. So we decided to request your feedback and discuss here about how we can improve our open source processes. Please write here, what do you dislike about the current development process, and your suggestions, how it can be improved. 

@farodin91
@Undin 
@kumbayo 
@alygin 
... and all the rest
... and even someone who just want to start contributing

**Comment by vlad20012** [comment]
Personally I have the following claim:
It's hard to know who work on what. It may be a problem if 2 guys do the same work, or if someone looking to work and want to ask "does not someone already work on this?". Currently I'm trying to ship the [Worklist](https://github.com/intellij-rust/intellij-rust/wiki/Worklist) - a wiki page open for editing for everyone, but I don't sure it will really work.

**Comment by farodin91** [comment]
I would like to have short term goals and long term goals.  For me, I'm looking the most time into features like to work on, or I miss from the IDE. I think clear goals could push the development speed. 
- Short term goals could be a monthly sprint.
- Long term goals could be the next version.

Getting the changelog into this week in rust would be cool.

Add a label for mentor issues. I would like to help.

**Comment by vlad20012** [comment]
@farodin91, thanks for the feedback!
1. Sounds reasonable...
2. We have a [weakly changelog](https://intellij-rust.github.io/thisweek/). What's wrong with it?)
3. Сould you reveal the idea of label? You're about something like [this](https://github.com/nodejs/node-v0.x-archive/issues/25618)?

**Comment by farodin91** [comment]
@vlad20012 
2. This week in rust has a part of blog entries, it would be cool to link the changelog into this section. To get more visibility.
3. I known from servo, but looks like the same as node is doing.

**Comment by vlad20012** [comment]
Ok, I get it.

**Comment by kumbayo** [comment]
When I work on something and it it not ready for inclusion yet I still try to push it as a "wip-xxx" branch in my fork of intellij-rust. Not sure if this helps others to see what i am playing around with.
But personally i have added all your forks to see what is currently being developed.

I now also listed my ideas for future/currently in progress work in the wiki page.

I also like the idea of some long term goals.
It might be interesting for less involved people to know what they can expect in the nearish future.

**Comment by farodin91** [comment]
@kumbayo Personally if i start working on a new feature, I'm trying to open a PR as early as possible to show development.

**Comment by alygin** [comment]
I agree that having some kind of a plan would be cool. But I think, it should be comprised of relatively short-term, _achievable_ goals that can be broken into several issues. I mean, it’s nice to come up with a plan with “Tests support”, “Type inference” and “Debugger” in it. But as practice [shows](https://github.com/intellij-rust/intellij-rust/issues/424), such plans tend to be too opaque to follow in long term, especially in an open source project. I, personally, would prefer seeing several meta-issues, like “Make Structure View useful” (fix known bugs, add filters, display macro defs etc.), “Improve build cfg panel” (auto detect channels, select target, gui elements for most used options etc), “Bring Rust settings panel in order” (separate settings into groups, give them proper names and shortcuts, etc), “Add all error annotations given the current resolver state”, etc. Something not too big, with clear boundaries and linked issues, something that delivers visible value in a particular area, not necessarily bound to timeline. So, even new people could start with closing easy issues and feel their contribution into something bigger, that will bear fruit in the nearest future. Closing issues is motivating, closing meta-issues is even more so.

Regarding big, long-term goals, they could be gathered in some memo along with general attitude to the plugin development. Alexey has already put great effort into telling developers about which way the plugin goes and why. But his explanations are scattered in many issues, chats and forums. I think, gathering a summary somewhere would be useful for new contributors and those who are interested in such plans. Something similar to this [FAQ](https://intellij-rust.github.io/docs/faq.html), but a bit more detailed and covering more questions (tests, error annotations, macros support, other things that people often ask), would be of great help, I think.
 
One more thing is labels. I suppose, a decent amount of labels would help to work with growing number of issues. For instance, to separate bugs from new features and enhancements, at least where it’s clear.

Unfortunately, I haven't had enough free time to contribute recently, but I hope to get back someday, use the “meta” label to find some cool meta-issue and grab an “easy”-labelled issue from it to pitch in :)

**Comment by vlad20012** [comment]
Thanks @alygin for such a detailed feedback!

I see you all suggest a better planning, so I think we definitely should introduce it somehow =)

About FAQ, we can create a wiki page and gradually fill it together. Ok, I even [created it](https://github.com/intellij-rust/intellij-rust/wiki/FAQ) :)

**Comment by vlad20012** [comment]
Ok, I think we can discuss here about _The Plan_. Rust community currently promotes the "impl period" for the rest of 2017, so I think we can formulate out goals for that "impl period" too.

I am not good in any planning (=D), so I can just describe my own goals (what I want to see in Intellij-Rust _right_now_ and what I'm working/planning to work on)
* Totally ultimate type inference
* Totally ultimate name resolution
  * macro expansion
* unresolved reference annotation
* auto import (based on all the above)

First, type inference is very close to completion. Yes, there probably will be a lot of bugs, but they will be classified as a _real_bugs_, not the "feature not implemented".
Then, name resolution (including macros) is a basic infrastructure tasks in the IDE at all. Everything depends on it (most relevant is navigation). Then, first consumer of the *full* name resolution is the _auto import_. I think no one doubts its necessity. Currently it's stalled on macros. Without macros support we will get many false positives with _unresolved reference annotation_ and it will be totally unusable. So, macros first.

So I think these goals are really most relevant for now.
And maybe I'm too optimistic, but I think these goals are achievable in 2017.
What do you think about this?

**Comment by mkaput** [comment]
To me they seem totally reasonable. Godspeed!

**Comment by farodin91** [comment]
I like to use the _impl period_ to get intellij rust in solid state:

> * Totally ultimate type inference
> * Totally ultimate name resolution
>   * macro expansion
> * unresolved reference annotation
> * auto import (based on all the above)

**Comment by farodin91** [comment]
A next step could be create a milestone impl period and add all correlated issues.

**Comment by farodin91** [comment]
We could also announce some issues in this week in rust.

**Comment by vlad20012** [comment]
@farodin91 :+1:, working on it

**Comment by vlad20012** [comment]
@farodin91 Hmm, there is preparing something new intead of "Call for Participation", much more useful. We will there =)

**Comment by Undin** [comment]
Sorry for the very slow response.
My wishes are already expressed by @alygin - it will be great to have a hierarchy in issue list where root issues are some meta-issues about some part of plugin and children are quite small tasks to implement/fix something related with the corresponding meta-issue. I think it can be easily implemented with labels.

**Comment by anton-okolelov** [comment]
> Please write here, what do you dislike about the current development process, and your suggestions, how it can be improved.

Is there any way to determine what issues is more important than other issues? Priority labels, maybe?

I mean, any project has too many issues to fix, but succesful projects concentrate on important things.  [Theory of constraints](https://en.wikipedia.org/wiki/Theory_of_constraints)  and so on

**Comment by Kobzol** [comment]
For me it would be nice if there was a way to talk about "meta" stuff with the maintainers outside of issues/PRs. There is the gitter site, but it seems to me that it's not very active. Maybe something like https://docs.github.com/en/discussions could be used?

**Comment by Kobzol** [comment]
To expand on my last comment, sometimes I want to ask or suggest a very general thing and I'm not sure where to do it. PRs are usually very narrowly focused on a single feature, issues do not seem to be the right fit, and the gitter site is not very active and sometimes it's hard to get an answer.

I like the concept of [GH discussions](https://docs.github.com/en/discussions) mainly because it's integrated directly into the repository. It therefore avoids the annoying problem of checking activity on an external site, which is needed with solutions like gitter, Discord, Slack, Zulip, etc.

**Comment by vlad20012** [comment]
@Kobzol I'll try to think about it again. Previously I decided that discussions don't help because they are almost the same as issues, but they're other entities, so they only increase the chaos. About the problem of checking activity, I'm not sure it helps since it's indeed another place to check, like "I should check PRs, I should check issues, I should check Discussions" (Not sure if I'm doing well even with the first point, haha).

**Comment by Kobzol** [comment]
Fair point that they are just another place to check, although much "closer" to issues and PRs than say, gitter.

They are indeed not that different from issues. I think that the main difference is mainly psychological. Some people (myself included) feel reluctant to open an "issue" to "just" discuss something, it feels like abusing a mechanism to do something it was not intended to do (although it's really just about the name - "issue").

Actually, I suggested discussions as a kind of an intermediate step. Personally, most of the time I would actually prefer to have a more "real-time" solution to contact the maintainers, like Slack/Zulip/Discord/chat/etc. Often I just want to "ping" something to make sure it's not forgotten, or just ask a "quick question" that might save me work.

But at the same time I feel slight anxiety because I worry that I will annoy/spam you. Also, it seems to me that there's already a lot of work for you and it's unrealistic to expect you to answer things on real-time chat in addition to all the work you're putting into the plugin.

**Comment by vlad20012** [comment]
Ok, let's [try it](https://github.com/intellij-rust/intellij-rust/discussions).

I agree on psychological differences with issues and agree that it will be a bit easier for us to track Discussions than Gitter

**Comment by Kobzol** [comment]
> Ok, let's try it.

You don't have to say that twice, I noticed it about an hour ago and already wrote [a lengthy post](https://github.com/intellij-rust/intellij-rust/discussions/7511) to try it out :D
